### PR TITLE
Implement deno.truncate

### PR DESCRIPTION
**Pull request by ztplz**
Node:  `fs.truncate(path[, len], callback)`
Go:  `func Truncate(path string, length int64) (err error)`
Python:  `os.truncate(path, length)`
Rust:  `pub fn set_len(&self, size: u64) -> Result<()>`

### CODE REVIEW

File: src/handlers.rs [hunk 1]
```diff
@@ -701,3 +702,21 @@ fn handle_symlink(i: *const isolate, base: &msg::Base) -> Box<Op> {
     }()))
   }
 }
+
+fn handle_truncate(i: *const isolate, base: &msg::Base) -> Box<Op> {
+  let deno = from_c(i);
+  if !deno.flags.allow_write {
+    return odd_future(permission_denied());
+  }
+   let msg = base.msg_as_truncate().unwrap();
```
Comment (kevinkassimo) [inline src/handlers.rs]: Remove this extra space (you might want to run `./tools/format.py` and `./tools/lint.py`)

File: src/handlers.rs [hunk 2]
```diff
@@ -701,3 +702,21 @@ fn handle_symlink(i: *const isolate, base: &msg::Base) -> Box<Op> {
...
+  let name = msg.name().unwrap();
+  let len = msg.len();
+  Box::new(futures::future::result(|| -> OpResult {
+    debug!("handle_mkdir {}", name);
+    debug!("handle_mkdir {}", len);
```
Comment (kevinkassimo) [inline src/handlers.rs]: I believe you could do
```rs
debug!("handle_mkdir {} {}", name, len);
```

Comment (ztplz) [reply src/handlers.rs]: Thanks you

File: js/truncate_test.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  const decoder = new TextDecoder("utf-8");
+  const text = decoder.decode(data);
+  return text;
+}
+
+testPerm({ write: true }, function truncateSyncSuccess() {
+  const enc = new TextEncoder();
+  const d = enc.encode("Hello");
+  const filename = deno.makeTempDirSync() + "/test_truncateSync.txt";
+  deno.writeFileSync(filename, d, 0o666);
```
Comment (ry) [inline js/truncate_test.ts]: Remove optional mode argument `0o666`, it's not relevant to this test.

File: js/truncate_test.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  deno.truncateSync(filename, 5);
+  data = readDataSync(filename);
+  assertEqual(data.length, 5);
+  // test len less than 0
+  deno.truncateSync(filename, -5);
+  data = readDataSync(filename);
+  assertEqual(data.length, 0);
+  
+  // delete test txt
+  deno.removeSync(filename);
```
Comment (ry) [inline js/truncate_test.ts]: I would remove all the comments here - it's self explanatory. Also remove blank links where appropriate.

File: src/handlers.rs [hunk 5]
```diff
@@ -701,3 +702,20 @@ fn handle_symlink(i: *const isolate, base: &msg::Base) -> Box<Op> {
...
+  let deno = from_c(i);
+  if !deno.flags.allow_write {
+    return odd_future(permission_denied());
+  }
+  let msg = base.msg_as_truncate().unwrap();
+  let name = msg.name().unwrap();
+  let len = msg.len();
+  Box::new(futures::future::result(|| -> OpResult {
+    debug!("handle_truncate {} {}", name, len);
+    let mut options = fs::OpenOptions::new();
```
Comment (ry) [inline src/handlers.rs]: What about using this https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.truncate

File: src/handlers.rs [hunk 6]
```diff
@@ -701,3 +702,20 @@ fn handle_symlink(i: *const isolate, base: &msg::Base) -> Box<Op> {
...
+    let f = options.write(true).open(name)?;
+    f.set_len(len as u64)?;
```
Comment (ry) [inline src/handlers.rs]: This is how you truncate in rust? This seems odd. 
Can you link me to documentation? I can't find anything on this....

File: src/handlers.rs (hunk 5 above)
Comment (kevinkassimo) [reply src/handlers.rs]: I believe that method would truncate the whole file to length 0

File: src/handlers.rs (hunk 6 above)
Comment (kevinkassimo) [reply src/handlers.rs]: I believe he meant this call: [`File.set_len`](https://doc.rust-lang.org/std/fs/struct.File.html#method.set_len)

Comment (ry) [reply src/handlers.rs]: Thanks

Comment (ztplz) [reply src/handlers.rs]: @ry  Here is link. [https://doc.rust-lang.org/std/fs/struct.File.html#method.set_len](https://doc.rust-lang.org/std/fs/struct.File.html#method.set_len)     In order for use `set_len`, write access must be used.

**Comment by ztplz** [comment]
@ry   Sure.

File: src/handlers.rs [hunk 7]
```diff
@@ -846,3 +847,23 @@ fn handle_read_link(
...
+  data: &'static mut [u8],
+) -> Box<Op> {
+  assert_eq!(data.len(), 0);
+  if !state.flags.allow_write {
+    return odd_future(permission_denied());
+  }
+  let msg = base.msg_as_truncate().unwrap();
+  let name = msg.name().unwrap();
+  let len = msg.len();
+  Box::new(futures::future::result(|| -> OpResult {
```
Comment (ry) [inline src/handlers.rs]: Please use `blocking!` here

Comment (ztplz) [reply src/handlers.rs]: @ry  Done.

**Pull request merged**